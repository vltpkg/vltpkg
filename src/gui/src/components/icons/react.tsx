export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
