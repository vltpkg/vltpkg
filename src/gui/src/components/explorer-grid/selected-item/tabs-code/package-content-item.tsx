import { useState } from 'react'
import {
  AlertDialog,
  AlertDialogTrigger,
  AlertDialogTitle,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogContent,
  AlertDialogDescription,
} from '@/components/ui/alert-dialog.tsx'
import { cn } from '@/lib/utils.ts'
import { formatDownloadSize } from '@/utils/format-download-size.ts'
import { getPackageContentIcon } from '@/components/explorer-grid/selected-item/tabs-code/utils.ts'
import {
  LARGE_FILE_ALERT,
  UNKNOWN_TYPE_ALERT,
} from '@/components/explorer-grid/selected-item/tabs-code/constants.ts'

import type {
  FsItemWithNone,
  ItemAlerts,
} from '@/components/explorer-grid/selected-item/tabs-code/types.ts'

export const PackageContentItem = ({
  item,
  onClick,
}: {
  item: FsItemWithNone
  onClick?: () => void
}) => {
  const [alert, setAlert] = useState<boolean>(false)
  const [alertContent, setAlertContent] = useState<ItemAlerts | null>(
    null,
  )
  const Icon = getPackageContentIcon(item.type)
  const FIVE_MB = 5 * 1024 * 1024

  const handleClick = () => {
    if (
      item.type === 'file' &&
      typeof item.size === 'number' &&
      item.size >= FIVE_MB
    ) {
      setAlertContent(LARGE_FILE_ALERT)
      setAlert(true)
      return
    }
    if (item.type === 'other') {
      setAlertContent(UNKNOWN_TYPE_ALERT)
      setAlert(true)
      return
    }
    onClick?.()
  }

  return (
    <AlertDialog open={alert} onOpenChange={setAlert}>
      <AlertDialogTrigger asChild onClick={e => e.preventDefault()}>
        <div
          role="link"
          className={cn(
            'group/package-content-item grid cursor-default grid-cols-12 bg-transparent px-6 py-2 transition-colors duration-100',
            'hover:bg-neutral-200',
            'hover:dark:bg-neutral-800',
          )}
          onClick={handleClick}>
          <div className="col-span-6 flex items-center gap-2">
            {Icon && (
              <div className="flex size-4 items-center justify-center">
                <Icon
                  className="size-4 text-muted-foreground"
                  strokeWidth={1}
                />
              </div>
            )}
            <p className="text-sm font-medium group-hover/package-content-item:underline">
              {item.name}
            </p>
          </div>
          <div className="col-span-3 flex justify-center">
            <p className="text-sm font-medium text-muted-foreground">
              {item.type !== 'none' && item.type}
            </p>
          </div>
          <div className="col-span-3 flex justify-end text-right">
            <p className="font-mono text-sm font-medium tabular-nums text-muted-foreground">
              {item.type !== 'none' ?
                item.size && item.size !== 0 ?
                  formatDownloadSize(item.size)
                : '-'
              : ''}
            </p>
          </div>
        </div>
      </AlertDialogTrigger>
      <AlertDialogContent>
        <AlertDialogHeader>
          <AlertDialogTitle>{alertContent?.title}</AlertDialogTitle>
          <AlertDialogDescription>
            {alertContent?.description}
          </AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel className="rounded-xl">
            Cancel
          </AlertDialogCancel>
          <AlertDialogAction
            className="rounded-xl"
            onClick={() => {
              setAlert(false)
              onClick?.()
            }}>
            Open
          </AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialog>
  )
}
