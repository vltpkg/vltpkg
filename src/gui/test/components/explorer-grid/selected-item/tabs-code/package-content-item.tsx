import { test, expect, afterEach, vi } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { PackageContentItem } from '@/components/explorer-grid/selected-item/tabs-code/package-content-item.tsx'

vi.mock('@/components/ui/alert-dialog.tsx', () => ({
  AlertDialog: 'gui-alert-dialog',
  AlertDialogTrigger: 'gui-alert-dialog-trigger',
  AlertDialogTitle: 'gui-alert-dialog-title',
  AlertDialogAction: 'gui-alert-dialog-action',
  AlertDialogCancel: 'gui-alert-dialog-cancel',
  AlertDialogFooter: 'gui-alert-dialog-footer',
  AlertDialogHeader: 'gui-alert-dialog-header',
  AlertDialogContent: 'gui-alert-dialog-content',
  AlertDialogDescription: 'gui-alert-dialog-description',
}))

vi.mock('lucide-react', () => ({
  File: 'gui-file-icon',
  FileText: 'gui-file-text-icon',
  Folder: 'gui-folder-icon',
  FolderSymlink: 'gui-folder-symlink-icon',
  FileQuestion: 'gui-file-question-icon',
}))

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  cleanup()
})

test('PackageContentItem renders directory row', () => {
  const { container } = render(
    <PackageContentItem
      item={{ name: 'src', type: 'directory', size: 0 }}
    />,
  )
  expect(container.innerHTML).toMatchSnapshot()
})

test('PackageContentItem renders file row with formatted size', () => {
  const { container } = render(
    <PackageContentItem
      item={{ name: 'file.ts', type: 'file', size: 1024 }}
    />,
  )
  expect(container.innerHTML).toMatchSnapshot()
})

test('PackageContentItem renders none row without icon and size', () => {
  const { container } = render(
    <PackageContentItem
      item={{ name: '../', type: 'none', size: 0 }}
    />,
  )
  expect(container.innerHTML).toMatchSnapshot()
})
