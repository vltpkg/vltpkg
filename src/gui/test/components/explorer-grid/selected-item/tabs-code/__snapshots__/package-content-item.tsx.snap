// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PackageContentItem renders directory row 1`] = `

<gui-alert-dialog open="false">
  <gui-alert-dialog-trigger aschild="true">
    <div
      role="link"
      class="group/package-content-item grid cursor-default grid-cols-12 bg-transparent px-6 py-2 transition-colors duration-100 hover:bg-neutral-200 hover:dark:bg-neutral-800"
    >
      <div class="col-span-6 flex items-center gap-2">
        <div class="flex size-4 items-center justify-center">
          <gui-folder-icon
            classname="size-4 text-muted-foreground"
            strokewidth="1"
          >
          </gui-folder-icon>
        </div>
        <p class="text-sm font-medium group-hover/package-content-item:underline">
          src
        </p>
      </div>
      <div class="col-span-3 flex justify-center">
        <p class="text-sm font-medium text-muted-foreground">
          directory
        </p>
      </div>
      <div class="col-span-3 flex justify-end text-right">
        <p class="font-mono text-sm font-medium tabular-nums text-muted-foreground">
          -
        </p>
      </div>
    </div>
  </gui-alert-dialog-trigger>
  <gui-alert-dialog-content>
    <gui-alert-dialog-header>
      <gui-alert-dialog-title>
      </gui-alert-dialog-title>
      <gui-alert-dialog-description>
      </gui-alert-dialog-description>
    </gui-alert-dialog-header>
    <gui-alert-dialog-footer>
      <gui-alert-dialog-cancel classname="rounded-xl">
        Cancel
      </gui-alert-dialog-cancel>
      <gui-alert-dialog-action classname="rounded-xl">
        Open
      </gui-alert-dialog-action>
    </gui-alert-dialog-footer>
  </gui-alert-dialog-content>
</gui-alert-dialog>

`;

exports[`PackageContentItem renders file row with formatted size 1`] = `

<gui-alert-dialog open="false">
  <gui-alert-dialog-trigger aschild="true">
    <div
      role="link"
      class="group/package-content-item grid cursor-default grid-cols-12 bg-transparent px-6 py-2 transition-colors duration-100 hover:bg-neutral-200 hover:dark:bg-neutral-800"
    >
      <div class="col-span-6 flex items-center gap-2">
        <div class="flex size-4 items-center justify-center">
          <gui-file-text-icon
            classname="size-4 text-muted-foreground"
            strokewidth="1"
          >
          </gui-file-text-icon>
        </div>
        <p class="text-sm font-medium group-hover/package-content-item:underline">
          file.ts
        </p>
      </div>
      <div class="col-span-3 flex justify-center">
        <p class="text-sm font-medium text-muted-foreground">
          file
        </p>
      </div>
      <div class="col-span-3 flex justify-end text-right">
        <p class="font-mono text-sm font-medium tabular-nums text-muted-foreground">
          1 KB
        </p>
      </div>
    </div>
  </gui-alert-dialog-trigger>
  <gui-alert-dialog-content>
    <gui-alert-dialog-header>
      <gui-alert-dialog-title>
      </gui-alert-dialog-title>
      <gui-alert-dialog-description>
      </gui-alert-dialog-description>
    </gui-alert-dialog-header>
    <gui-alert-dialog-footer>
      <gui-alert-dialog-cancel classname="rounded-xl">
        Cancel
      </gui-alert-dialog-cancel>
      <gui-alert-dialog-action classname="rounded-xl">
        Open
      </gui-alert-dialog-action>
    </gui-alert-dialog-footer>
  </gui-alert-dialog-content>
</gui-alert-dialog>

`;

exports[`PackageContentItem renders none row without icon and size 1`] = `

<gui-alert-dialog open="false">
  <gui-alert-dialog-trigger aschild="true">
    <div
      role="link"
      class="group/package-content-item grid cursor-default grid-cols-12 bg-transparent px-6 py-2 transition-colors duration-100 hover:bg-neutral-200 hover:dark:bg-neutral-800"
    >
      <div class="col-span-6 flex items-center gap-2">
        <p class="text-sm font-medium group-hover/package-content-item:underline">
          ../
        </p>
      </div>
      <div class="col-span-3 flex justify-center">
        <p class="text-sm font-medium text-muted-foreground">
        </p>
      </div>
      <div class="col-span-3 flex justify-end text-right">
        <p class="font-mono text-sm font-medium tabular-nums text-muted-foreground">
        </p>
      </div>
    </div>
  </gui-alert-dialog-trigger>
  <gui-alert-dialog-content>
    <gui-alert-dialog-header>
      <gui-alert-dialog-title>
      </gui-alert-dialog-title>
      <gui-alert-dialog-description>
      </gui-alert-dialog-description>
    </gui-alert-dialog-header>
    <gui-alert-dialog-footer>
      <gui-alert-dialog-cancel classname="rounded-xl">
        Cancel
      </gui-alert-dialog-cancel>
      <gui-alert-dialog-action classname="rounded-xl">
        Open
      </gui-alert-dialog-action>
    </gui-alert-dialog-footer>
  </gui-alert-dialog-content>
</gui-alert-dialog>

`;
